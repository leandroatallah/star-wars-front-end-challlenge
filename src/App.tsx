import { Helmet, HelmetProvider } from 'react-helmet-async'
import { ThemeProvider, createGlobalStyle } from 'styled-components'
import { connect } from 'react-redux'
import './app.css'
import Welcome from './components/welcome'
import MasterResults from './components/master-results'
import themeColors from './config/theme-colors'

const GlobalStyle = createGlobalStyle`
  * {
      margin: 0;
      padding: 0;
      border: 0;
      vertical-align: baseline;
  }

  body {
    font-family: 'Montserrat', sans-serif;
  }
`

type ThemeState = {
  theme: string
}

type AppProps = {
  theme: string
}

const App = (props: AppProps): JSX.Element => {
  return (
    <HelmetProvider>
      <div className="App">
        <Helmet>
          <meta charSet="utf-8" />
          <title>iClinic Frontend Challenge</title>
          <link rel="preconnect" href="https://fonts.googleapis.com" />
          <link rel="preconnect" href="https://fonts.gstatic.com" crossOrigin="true" />
          <link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@400;700&display=swap" rel="stylesheet" />
        </Helmet>
        <ThemeProvider theme={themeColors}>
          <GlobalStyle />
          {props.theme === '' ? <Welcome /> : <MasterResults />}
        </ThemeProvider>
      </div>
    </HelmetProvider>
  )
}

const mapStateToProps = (state: ThemeState) => {
  return {
    theme: state.theme,
  }
}

export default connect(mapStateToProps)(App)
