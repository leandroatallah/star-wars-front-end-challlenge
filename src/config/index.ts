export const CHANGE_THEME = 'CHANGE_THEME'
export const THEME_DARK_SIDE = 'THEME_DARK_SIDE'
export const THEME_LIGHT_SIDE = 'THEME_LIGHT_SIDE'
