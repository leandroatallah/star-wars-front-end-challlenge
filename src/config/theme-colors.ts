type myTheme = {
    colors: {
        blue: string
        yellow: string
        black: string
        white: string
    }
}

const themeColors: myTheme = {
    colors: {
        blue: '#2C97D1',
        yellow: '#FBFE63',
        black: '#2A2A2A',
        white: '#FFF',
    },
}

export default themeColors