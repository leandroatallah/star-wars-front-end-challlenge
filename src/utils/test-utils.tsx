import { Provider } from 'react-redux'
import { render } from '@testing-library/react'
import { CHANGE_THEME } from '../config'
import { createStore, Store } from 'redux'

const startingState = { theme: '' }
function reducer(state = startingState, action: { type: string; payload: string }) {
    switch (action.type) {
        case CHANGE_THEME:
            return { ...state, theme: action.payload }
        default:
            return state
    }
}

type storeProps = {
    initialState?: {
        theme: string
    },
    store?: Store<{
        theme: string;
    }, {
        type: string;
        payload: string;
    }>
}

export function renderWithRedux(
    component: JSX.Element,
    { initialState, store = createStore(reducer, initialState) }: storeProps = {},
) {
    return {
        ...render(<Provider store={store} > {component} </Provider>),
    }
}
