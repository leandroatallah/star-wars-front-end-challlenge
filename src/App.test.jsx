import '@testing-library/jest-dom'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'
import App from './App'
import themeColors from './config/theme-colors'
import { THEME_DARK_SIDE, THEME_LIGHT_SIDE } from './config'
import { cleanup, fireEvent } from '@testing-library/react'
import { createStore } from 'redux'
import { renderWithRedux } from './utils/test-utils'

afterEach(cleanup)

describe('<App />', () => {
    it('Render with redux', () => {
        const { container } = renderWithRedux(<App />)
        expect(container).toMatchSnapshot()
    })

    it('should render the Welcome and not render MasterResults by default', () => {
        const { getByTestId } = renderWithRedux(<App />)
        expect(getByTestId('welcome')).toBeInTheDocument()
        expect(() => getByTestId('master-results')).toThrow()
    })

    it('should render MasterResults and not render Welcome when dark theme is setted', () => {
        const store = createStore(() => ({ theme: THEME_DARK_SIDE }))
        const { getByTestId } = renderWithRedux(<App />, { store })
        expect(getByTestId('master-results')).toBeInTheDocument()
        expect(() => getByTestId('welcome')).toThrow()
    })

    it('should render MasterResults and not render Welcome when light theme is setted', () => {
        const store = createStore(() => ({ theme: THEME_LIGHT_SIDE }))
        const { getByTestId } = renderWithRedux(<App />, { store })
        expect(getByTestId('master-results')).toBeInTheDocument()
        expect(() => getByTestId('welcome')).toThrow()
    })

    it('should fetch swapi and show light side master if get light side result', () => {
        const { getByTestId } = renderWithRedux(
            <ThemeProvider theme={themeColors}>
                <App />
            </ThemeProvider>,
        )
        const welcome = getByTestId('welcome')
        const button = welcome.querySelector('button')
        fireEvent.click(button)
    })

    // it('should fetch swapi and show dark side master if get dark side result', () => {
    //     const { getByTestId, getByRole } = renderWithRedux(
    //         <ThemeProvider theme={themeColors}>
    //             <App />
    //         </ThemeProvider>,
    //     )
    //     const button = getByRole('button')
    //     fireEvent.click(button)
    //     // expect()
    // })

    // it('should fetch swapi and show warning if get no api result', () => {
    //     const { getByTestId, getByRole } = renderWithRedux(
    //         <ThemeProvider theme={themeColors}>
    //             <App />
    //         </ThemeProvider>,
    //     )
    //     const button = getByRole('button')
    //     fireEvent.click(button)
    //     expect(getByTestId('warning')).toBeInTheDocument()
    // })
})
