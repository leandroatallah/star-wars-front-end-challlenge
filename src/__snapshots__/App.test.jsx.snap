// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Render with redux 1`] = `
.c1 {
  background-color: #2C97D1;
  display: inline-block;
  height: 56px;
  font-size: 18px;
  line-height: 1.218;
  font-weight: 700;
  padding: 0 37px;
  border-radius: 10px;
  color: #FFF;
  cursor: pointer;
}

.c0 {
  margin-top: 210px;
  text-align: center;
  color: #2C97D1;
}

.c0 h1 {
  font-size: 72px;
  line-height: 1.2;
  margin-bottom: 7px;
  font-weight: 400;
}

.c0 h1 strong {
  font-weight: 700;
}

.c0 h2 {
  font-size: 14px;
  line-height: 1.218;
  font-weight: 400;
  -webkit-letter-spacing: calc(14px * 0.35);
  -moz-letter-spacing: calc(14px * 0.35);
  -ms-letter-spacing: calc(14px * 0.35);
  letter-spacing: calc(14px * 0.35);
  text-transform: uppercase;
  margin-bottom: 162px;
}

<div>
   
  <div
    class="App"
  >
    <section
      class="c0"
      data-testid="welcome"
    >
      <h1>
        Welcome to 
        <strong>
          iClinic
        </strong>
      </h1>
      <h2>
        Frontend Challenge
      </h2>
      <button
        class="c1"
        data-testid="button"
        type="button"
      >
        S T A R T
      </button>
    </section>
  </div>
   
</div>
`;
